core/src/test/java/org/elasticsearch/indices/recovery/RecoveryTests.java

```diff
@@ -108,12 +114,21 @@ public void testRecoveryWithOutOfOrderDelete() throws Exception {
                 SourceToParse.source(orgReplica.shardId().getIndexName(), "type", "id", new BytesArray("{}"), XContentType.JSON),
                 u -> {});
 
+            if (isolateFirstBatch) {
```

Comment:
we discussed an alternative simpler simulation.

Reply:
@bleskes, This test is more fun than we thought. Both `delete #1` and `index #0` are either kept or deleted at the same time.

1. To delete `delete #1` but keep `index #0`, we need isolate `delete #1` in its own translog `gen #1` (by flushing or rolling).

2. To keep `index #0` longer than `delete #1`, we need to store `index #0` with a higher seqno operation. However, this operation increases `max_seq`, then prevents the `delete 
 #0` from being deleted as we still have to keep the old commit (it's the only safe commit).

I used the size retention policy to keep `index #0` longer.

Reply:
How about 
- delete #1 
- roll generation  (to create gen 2)
- index #0 
- index #3 
- flush (commit point has max seq no 3, and local checkpoint 1 -> points at gen 2, previous commit point is maintained)
- index #2 
- index #5 
- update the global checkpoint to #3
- tell lucene to delete unneeded commits so it will remove the old commit  (via IndexWriter#deleteUnusedFiles). Once we do a rollback on recovery, we can just flush again.

I think this will work?

Reply:
Yep, this works. I pushed https://github.com/elastic/elasticsearch/pull/27606/commits/6733b750d714955d2715eb1d67f8e31305e48b9e